### PR TITLE
drivers: pwm: nrf_sw: support generating PWM based on RTC

### PR DESCRIPTION
**Pull request by mniestroj**
So far nRF's TIMER was used for generating PWM signal. Add support for
generating PWM based on RTC, which is sourced by 32KHz low frequency
crystal. This allows to use low frequency PWM with much lower power
consumption, because high frequency clock path can be disabled.

Tested with RTC0 and RTC2, with `channel-count` equal 1 or 2.

### CODE REVIEW

File: dts/bindings/pwm/nordic,nrf-sw-pwm.yaml [hunk 1]
```diff
@@ -5,10 +5,15 @@ compatible: "nordic,nrf-sw-pwm"
 include: [pwm-controller.yaml, base.yaml]
 
 properties:
+    rtc-instance:
+      type: int
+      description: RTC instance to use for generating the PWM output signals
```
Comment (anangl) [inline dts/bindings/pwm/nordic,nrf-sw-pwm.yaml:10]: Descriptions of `channel-count`: https://github.com/zephyrproject-rtos/zephyr/blob/99f5fc4d21572ff3ff177a36683fb339364ec378/dts/bindings/pwm/nordic%2Cnrf-sw-pwm.yaml#L20
and `clock-prescaler`: https://github.com/zephyrproject-rtos/zephyr/blob/99f5fc4d21572ff3ff177a36683fb339364ec378/dts/bindings/pwm/nordic%2Cnrf-sw-pwm.yaml#L25
should be also updated.

**Review by anangl — CHANGES_REQUESTED** [review]
The following description needs to be also updated: https://github.com/zephyrproject-rtos/zephyr/blob/99f5fc4d21572ff3ff177a36683fb339364ec378/drivers/pwm/Kconfig.nrf5_sw#L12-L13
And [this definition](https://github.com/zephyrproject-rtos/hal_nordic/blob/fc12f5e2b47cf143254234c05f9ecbdb603a7101/nrfx_glue.h#L260-L262) as well:
```
#define NRFX_PPI_CHANNELS_USED_BY_PWM_SW \
    (BIT_MASK(DT_PROP(PWM_NRF5_SW_NODE, channel_count) * 2) \
         << DT_PROP(PWM_NRF5_SW_NODE, ppi_base))
```

**Comment by mniestroj** [comment]
```
-:51: ERROR:SPACING: need consistent spacing around '*' (ctx:WxV)
#51: FILE: drivers/pwm/pwm_nrf5_sw.c:41:
+	NRF_RTC_Type *rtc;
```
seems like a false positive. Any idea why checkpatch is complaining? After changing to `NRF_RTC_Type * rtc;` it stops warning us, but that is not the style used in Zephyr.

File: drivers/pwm/pwm_nrf5_sw.c [hunk 2]
```diff
@@ -160,31 +199,58 @@ static int pwm_nrf5_sw_pin_set(const struct device *dev, uint32_t pwm,
 	/* calc div, to scale down to fit in 16 bits */
 	div = period_cycles >> 16;
```
Comment (anangl) [inline drivers/pwm/pwm_nrf5_sw.c:200]: RTC has 24-bit compare registers.
But this "scaling" is anyway incorrect, even for the 16-bit TIMER case. Instead, something like this should be done above (for RTC, the TIMER case can be corrected in a separate PR):
```
if (USE_RTC && period_cycles > BITMASK(24)) {
	LOG_ERR("Too long period (%u)", period_cycles);
	return -EINVAL;
}
```

**Comment by anangl** [comment]
> ```
> -:51: ERROR:SPACING: need consistent spacing around '*' (ctx:WxV)
> #51: FILE: drivers/pwm/pwm_nrf5_sw.c:41:
> +	NRF_RTC_Type *rtc;
> ```
> 
> seems like a false positive. Any idea why checkpatch is complaining? After changing to `NRF_RTC_Type * rtc;` it stops warning us, but that is not the style used in Zephyr.

Looks like a false positive, indeed.

File: drivers/pwm/Kconfig.nrf5_sw [hunk 3]
```diff
@@ -9,5 +9,9 @@ config PWM_NRF5_SW
 	depends on SOC_FAMILY_NRF
 	help
 	  Enable driver to utilize PWM on the Nordic Semiconductor nRF5x series.
-	  This implementation provides up to 3 pins using one HF timer, two PPI
+
+	  When used with HF timer it provides up to 3 pins, utilizing two PPI
+	  channels per pin and one GPIOTE config per pin.
+
+	  When used with RTC it provides up to 2 pins, utilizing three PPI
```
[the remark this answers is not in the document]
Comment (mniestroj) [reply drivers/pwm/Kconfig.nrf5_sw:16]: My mistake, I have confused PPI with RTC's CC. It should be 3 of course.

File: drivers/pwm/pwm_nrf5_sw.c [hunk 4]
```diff
@@ -160,31 +199,58 @@ static int pwm_nrf5_sw_pin_set(const struct device *dev, uint32_t pwm,
...
+		NRF_PPI->CH[ppi_index].TEP =
+			(uint32_t) &(NRF_GPIOTE->TASKS_OUT[channel]);
+		NRF_PPI->CH[ppi_index + 1].EEP =
+			(uint32_t) &(rtc->EVENTS_COMPARE[config->map_size]);
+		NRF_PPI->CH[ppi_index + 1].TEP =
+			(uint32_t) &(NRF_GPIOTE->TASKS_OUT[channel]);
+		NRF_PPI->CH[ppi_index + 2].EEP =
+			(uint32_t) &(rtc->EVENTS_COMPARE[config->map_size]);
+		NRF_PPI->CH[ppi_index + 2].TEP =
+			(uint32_t) &(rtc->TASKS_CLEAR);
```
[the remark this answers is not in the document]
Comment (mniestroj) [reply drivers/pwm/pwm_nrf5_sw.c:235]: Seems reasonable to drop prescaling support, I cannot imagine use case for >512s PWM period.

So I guess we should two things:
1) force `clock-prescaler = <0>;` with `BUILD_ASSERT()`,
2) drop `div = period_cycles >> 16;` calculation entirely for RTC (the same can be done for TIMER) and just check that cycles value fit in 24 bits.

File: drivers/pwm/pwm_nrf5_sw.c (hunk 2 above)
Comment (mniestroj) [reply drivers/pwm/pwm_nrf5_sw.c:200]: Am I right then, that this scaling (`div = period_cycles >> 16;`) should be dropped entirely for both RTC and TIMER?

Comment (anangl) [reply drivers/pwm/pwm_nrf5_sw.c:200]: I think so. Currently it just uses incorrect settings, and thus gives incorrect PWM output. It will be better to return an error with some hint to the user, like in the mcux driver: https://github.com/zephyrproject-rtos/zephyr/blob/3c79b565fe7e6afed3687d95dd3e70bb93699a88/drivers/pwm/pwm_mcux.c#L58-L64

File: drivers/pwm/pwm_nrf5_sw.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (mniestroj) [reply drivers/pwm/pwm_nrf5_sw.c:248]: I've added commit which drops `div = period_cycles >> 16;` for TIMER. It just seemed more logical to drop broken mechanism, even in this PR, before introducing new features.

Additionally I've adjusted period_cycles and pulse_cycles to take one LFCLK cycle delay:
```
	rtc->CC[channel] = pulse_cycles - 1;
	rtc->CC[config->map_size] = period_cycles - 1;
```

File: drivers/pwm/pwm_nrf5_sw.c (hunk 2 above)
Comment (mniestroj) [reply drivers/pwm/pwm_nrf5_sw.c]: Added similar check for both TIMER and RTC.

**Review by anangl — APPROVED** [review]

File: dts/bindings/pwm/nordic,nrf-sw-pwm.yaml [hunk 5]
```diff
@@ -5,19 +5,32 @@ compatible: "nordic,nrf-sw-pwm"
 include: [pwm-controller.yaml, base.yaml]
 
 properties:
+    rtc-instance:
+      type: int
```
Comment (mbolivar-nordic) [inline dts/bindings/pwm/nordic,nrf-sw-pwm.yaml:9]: Why not phandle and pointing to the `nordic,nrf-rtc` directly? I think it should probably be done that way, just looking at this quickly.

File: dts/bindings/pwm/nordic,nrf-sw-pwm.yaml [hunk 6]
```diff
@@ -5,19 +5,32 @@ compatible: "nordic,nrf-sw-pwm"
...
+      description: RTC instance to use for generating the PWM output signals
+      required: false
+
     timer-instance:
       type: int
```
Comment (mbolivar-nordic) [inline dts/bindings/pwm/nordic,nrf-sw-pwm.yaml:14]: I see that this already uses an instance number, but I think it would be more appropriate to use phandles.

I haven't reviewed the code in detail, though. I am not putting a -1 or a +1 as a result, just offering a quick comment. I will review further when I can, just wanted to have something to turn around for the time zone difference.

Comment (mniestroj) [reply dts/bindings/pwm/nordic,nrf-sw-pwm.yaml:14]: I agree, that using phandles is better. The reason why I went with `rtc-instance` is that I wanted to solve one issue at a time, which is adding RTC support to generate PWM. If such conversion to phandle is straightforward to do and you can point to some other driver which resolves RTC/TIMER instance number based on phandle, then I can update this PR. If that is not so simple, then I prefer to convert `rtc-instance` (`timer-instance`) integer to `rtc` (`timer`) phandle in another PR.

**Comment by mbolivar-nordic** [comment]
Marcin Niestroj <notifications@github.com> writes:

> @mniestroj commented on this pull request.
>
>
>
>> @@ -5,19 +5,32 @@ compatible: "nordic,nrf-sw-pwm"
>  include: [pwm-controller.yaml, base.yaml]
>  
>  properties:
> +    rtc-instance:
> +      type: int
> +      description: RTC instance to use for generating the PWM output signals
> +      required: false
> +
>      timer-instance:
>        type: int
>
> I agree, that using phandles is better. The reason why I went with
> `rtc-instance` is that I wanted to solve one issue at a time, which is
> adding RTC support to generate PWM. If such conversion to phandle is
> straightforward to do and you can point to some other driver which
> resolves RTC/TIMER instance number based on phandle, then I can update
> this PR. If that is not so simple, then I prefer to convert
> `rtc-instance` (`timer-instance`) integer to `rtc` (`timer`) phandle
> in another PR.

I wasn't able to think of anything so far, so let's just do it your way.

**Comment by pabigot** [comment]
> I wasn't able to think of anything so far, so let's just do it your way.

I had a little time left for the month and also want to get rid of instance numbers so I gave it a shot.

A potential solution is in the top commit [here](https://github.com/pabigot/zephyr/commits/pr/29094).  It seems pretty clean, but the fact the Nordic RTC and TIMER devicetree nodes **still** don't provide the number of CC registers associated with the instance requires eliminating a compile-time check.

I've done build and value tests of these on samples/boards/bbc_microbit/sound, but I can't confirm the driver actually works because I don't have a buzzer to attach.

You're welcome to integrate this or not, as you wish.

**Comment by mniestroj** [comment]
@pabigot I really like your proposal. That way `rtc-instance` doesn't have to be introduced at all. Let's wait for `cc-num` PR and I will update this PR then.

**Comment by mniestroj** [comment]
There is one additional commit, that switches from timer instance number to timer phandle. Adding RTC support got simplified a little bit, so we got less preprocessor usage at the beginning of file.

File: drivers/pwm/pwm_nrf5_sw.c [hunk 7]
```diff
@@ -153,38 +205,59 @@ static int pwm_nrf5_sw_pin_set(const struct device *dev, uint32_t pwm,
...
-	div = period_cycles >> 16;
+	/* configure RTC / TIMER */
+	if (USE_RTC) {
+		rtc->EVENTS_COMPARE[channel] = 0;
+		rtc->EVENTS_COMPARE[config->map_size] = 0;
 
-	timer->EVENTS_COMPARE[channel] = 0;
-	timer->EVENTS_COMPARE[config->map_size] = 0;
+		rtc->CC[channel] = pulse_cycles - 1;
+		rtc->CC[config->map_size] = period_cycles - 1;
```
Comment (anangl) [inline drivers/pwm/pwm_nrf5_sw.c:214]: Maybe it would be good to add a comment why these `- 1` are needed here.

File: drivers/pwm/pwm_nrf5_sw.c [hunk 8]
```diff
@@ -15,17 +15,30 @@
...
 #error "Invalid number of PWM channels configured."
 #endif
 #define PWM_0_MAP_SIZE DT_INST_PROP(0, channel_count)
 
 struct pwm_config {
+#if USE_RTC
+	NRF_RTC_Type *rtc;
+#else
 	NRF_TIMER_Type *timer;
+#endif
```
Comment (anangl) [inline drivers/pwm/pwm_nrf5_sw.c:41]: How about using a union here? Maybe this could please the compliance check?

Comment (mniestroj) [reply drivers/pwm/pwm_nrf5_sw.c]: Indeed, using `union` helped with `checkpatch.pl`. `#ifdef` might have confused it somehow.

File: drivers/pwm/pwm_nrf5_sw.c (hunk 7 above)
Comment (mniestroj) [reply drivers/pwm/pwm_nrf5_sw.c:218]: Added a comment about it.

Comment (mniestroj) [reply drivers/pwm/pwm_nrf5_sw.c:218]: Added a comment about it.

File: dts/bindings/pwm/nordic,nrf-sw-pwm.yaml [hunk 9]
```diff
@@ -5,9 +5,9 @@ compatible: "nordic,nrf-sw-pwm"
 include: [pwm-controller.yaml, base.yaml]
 
 properties:
-    timer-instance:
-      type: int
-      description: Timer instance to use for generating the PWM output signals
+    generator:
+      type: phandle
+      description: Reference to TIMER instance for generating PWM output signals
```
Comment (mbolivar-nordic) [inline dts/bindings/pwm/nordic,nrf-sw-pwm.yaml]: This will break compatibility with any code that's using timer-instance.

Do you think you could fall back on `timer-instance` if `generator` is not defined?

Something like this:

```c
#if DT_INST_NODE_HAS_PROP(0, timer_instance) && DT_INST_NODE_HAS_PROP(0, generator)
#error "Please define either the timer-instance or generator property, but not both"
#elif DT_INST_NODE_HAS_PROP(0, timer_instance)
#define GENERATOR_ADDR _CONCAT(NRF_TIMER, DT_INST_PROP(0, timer_instance))
#else
#define GENERATOR_ADDR	((NRF_TIMER_Type *) DT_REG_ADDR(GENERATOR_NODE))
#endif
```

If you can put in a comment saying 'timer-instance' is deprecated, that would be useful.

Comment (pabigot) [reply dts/bindings/pwm/nordic,nrf-sw-pwm.yaml]: Didn't we just get support for deprecated properties from #29417?  We could support both for a while, right?

If `timer-instance` is removed from the binding I wouldn't expect the conditional code to work.

Comment (mbolivar-nordic) [reply dts/bindings/pwm/nordic,nrf-sw-pwm.yaml]: Sorry if I was unclear.

I am suggesting we do **not** remove this from the binding, exactly so the conditional code will work.

Comment (mniestroj) [reply dts/bindings/pwm/nordic,nrf-sw-pwm.yaml]: I have updated code to keep `timer-instance` as deprecated property. Thanks for all suggestions!

**Review by mbolivar-nordic — APPROVED** [review]
Nice work!

**Comment by carlescufi** [comment]
@mniestroj please update to SHA

File: west.yml [hunk 10]
```diff
@@ -56,7 +56,7 @@ manifest:
       revision: f1fa8241f8786198ba41155413243de36ed878a5
       path: modules/hal/infineon
     - name: hal_nordic
-      revision: 7430036d4f18cdf1f49158e97e5d57aacb44fb92
+      revision: pull/61/head
```
Comment (anangl) [inline west.yml:59]: @mniestroj Please update to:
```suggestion
      revision: 031065e8f9066b4e5fcda57f390af1cd67f35a65
```
as https://github.com/zephyrproject-rtos/hal_nordic/pull/61 got merged.
